<a href="https://colab.research.google.com/github/neilsambhu/carla-simulator2/blob/main/carla_simulator2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Sandbox for CARLA simulator following https://github.com/MichaelBosello/carla-colab. Original *.ipynb file: https://colab.research.google.com/github/MichaelBosello/carla-colab/blob/master/carla-simulator.ipynb. 

In case of poor functionality: Runtime > Restart runtime (Ctrl+M.)

In case of very poor functionality: Runtime > Factory reset runtime

2/17/2022 8:49 PM: instead of (1) "!" to execute shell commands, use (2) the Python "subprocess" module to ensure compatibility between iPython and Python (complete as of 2/17/2022 10:16 PM).

2/17/2022 10:23 PM: I made the Call_SubprocessPopen(cmd) function for easily calling shell commands. 

2/17/2022 11:02 PM: The subprocess.run will fail for command-line arguments separated by spaces. The subprocess.run requires a list.

In [1]:
import subprocess, sys
from sys import platform
def Call_SubprocessPopen(cmd):
  p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
  # p = subprocess.run(cmd, capture_output=True, text=True, check=True)
  for line in p.stdout.readlines():
    sys.stdout.write(line.decode("utf-8"))
    # sys.stdout.write((r"utf-8")line)
    pass
  # print(p.stdout)
if platform == "linux" or platform == "linux2":
  # for multi-line output: use "pip --help" instead of "pwd" 
  cmd = "pwd"
  # cmd = "pip --help"
  Call_SubprocessPopen(cmd)
elif platform == "win32":
  raise Exception("2/17/2022 8:59 PM: This code is designed to run in a Linux-based environment.")

/content


In [2]:
# <10 seconds
Call_SubprocessPopen("pip install git+https://github.com/demotomohiro/remocolab.git")

  Cloning https://github.com/demotomohiro/remocolab.git to /tmp/pip-req-build-5j663vx3
  Running command git clone -q https://github.com/demotomohiro/remocolab.git /tmp/pip-req-build-5j663vx3
  Created wheel for remocolab.py: filename=remocolab.py-0.1-py3-none-any.whl size=7553 sha256=0dc9681c31ad5858d42af344271efe38f07ebb1552ca53ef435f2a860d66542f
  Stored in directory: /tmp/pip-ephem-wheel-cache-qmbk2pud/wheels/80/37/44/386e71ab0cd5aef5cefdb1926b710a42fdb4fe2d07c3b1b08f
  Created wheel for pyngrok: filename=pyngrok-5.1.0-py3-none-any.whl size=19007 sha256=ed8ab00f8ee5a2d35c0b7253d57b85b729e2423d470d76d16e837b921edeaf61
  Stored in directory: /root/.cache/pip/wheels/bf/e6/af/ccf6598ecefecd44104069371795cb9b3afbcd16987f6ccfb3
Successfully built remocolab.py pyngrok


In [3]:
# <5.5 minutes
import remocolab
remocolab.setupVNC()

Warning! GPU of your assigned virtual machine is Tesla K80.
You might get better GPU by reseting the runtime.
Do you want to continue? [y/n] y
---
As ngrok doesn't work on colab for long time, default tunnel method has been changed to Argo tunnel.
Please read this for more details:
https://github.com/demotomohiro/remocolab/blob/master/README.md


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install openssh-server


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install xfce4
Install xfce4-terminal
ED25519 key fingerprint of host:
256 SHA256:7lPdGDzBgprm1umSb2F8J8JJsfItGUvsnR4iKAJOSno root@37ff38319ba1 (ED25519)
+--[ED25519 256]--+
|         . .     |
|        o . o    |
|       + o o .   |
|.o    = *   +    |
|B    + XSO o =   |
|+.E . +.^ O + .  |
| o . . =.O +     |
|      o.+ .      |
|       +o.       |
+----[SHA256]-----+

✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
root password: CMXUnfvQCgGFKxWxdA9rkViPA0h6jWRmRWDF0TmTgW8
colab password: 5U-qihVeewuhdZY8dXncXJ6mJFDp--qbaeFvPQLcNPw
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
---
Execute following command on your local machine and login before running TurboVNC viewer:
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@anger-write-acid-veteran.trycloudflare.com
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
VNC password: vvmsRzA0
VNC view on

2/20/2022 9:01 PM: Problem:
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
/bin/bash: line 0: exec: cloudflared: not found
kex_exchange_identification: Connection closed by remote host

[install cloudflared (https://pkg.cloudflare.com/#ubuntu-title) in Windows Subsystem for Linux (WSL) Ubuntu 20.04 LTS]

2/20/2022 9:08 PM: Problem
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -o ProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@vat-bracket-mean-hobbies.trycloudflare.com
/bin/bash: line 0: exec: cloudflared: not found
kex_exchange_identification: Connection closed by remote host

[WSL: sudo apt install cloudflared]

In [4]:
# <4 minutes

# def IOPub_DataRate(iopub_data_rate_limit=1000000.0, rate_limit_window=3.0):
#   import os
#   # os.environ['NotebookApp.iopub_data_rate_limit']=str(iopub_data_rate_limit)
#   Call_SubprocessPopen("--NotebookApp.iopub_data_rate_limit {}".format(iopub_data_rate_limit))
#   os.environ['NotebookApp.rate_limit_window']=str(rate_limit_window)
# # increase IOPub data rate; set the config variable `--NotebookApp.iopub_data_rate_limit`.
# # Current values:
# # NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
# # NotebookApp.rate_limit_window=3.0 (secs)
# IOPub_DataRate(1e7)

# Download CARLA
# Call_SubprocessPopen("sudo -u colab wget -O /home/colab/CARLA.tar.gz https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/CARLA_0.9.13.tar.gz")
# 2/22/2022 3:23 PM: try CARLA 0.9.10 instead of 0.9.13
Call_SubprocessPopen("sudo -u colab wget -O /home/colab/CARLA.tar.gz https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/CARLA_0.9.10.tar.gz")

# reset IOPub data rate
# IOPub_DataRate()

Streaming output truncated to the last 5000 lines.
1433550K .......... .......... .......... .......... .......... 37%  169M 3m29s
1433600K .......... .......... .......... .......... .......... 37% 10.7M 3m29s
1433650K .......... .......... .......... .......... .......... 37% 4.42M 3m29s
1433700K .......... .......... .......... .......... .......... 37% 19.9M 3m29s
1433750K .......... .......... .......... .......... .......... 37% 16.6M 3m29s
1433800K .......... .......... .......... .......... .......... 37% 5.42M 3m29s
1433850K .......... .......... .......... .......... .......... 37% 52.0M 3m29s
1433900K .......... .......... .......... .......... .......... 37% 10.4M 3m29s
1433950K .......... .......... .......... .......... .......... 37% 10.9M 3m29s
1434000K .......... .......... .......... .......... .......... 37%  128M 3m29s
1434050K .......... .......... .......... .......... .......... 37% 4.21M 3m29s
1434100K .......... .......... .......... .......... .......... 37% 1

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [5]:
# <3.5 minutes
Call_SubprocessPopen("sudo -u colab mkdir /home/colab/carla")
Call_SubprocessPopen("sudo -u colab tar -xf /home/colab/CARLA.tar.gz -C /home/colab/carla")

2/20/2022 10:41 PM: 

Disk is almost full

Your disk is almost full. Completely filling the disk will result in errors. Currently, 63.39 GB / 78.19 GB is being used. You can delete files in order to free up disk space. If you are interested in Colab VMs with larger disks, you may want to check out Colab Pro.

Setup PythonAPI

In [6]:
# <2 seconds
Call_SubprocessPopen("sudo -u colab rm -rf /home/colab/carla/PythonAPI/carla/dist/*.egg")
Call_SubprocessPopen("sudo -u colab wget https://github.com/MichaelBosello/carla-colab/raw/master/carla-0.9.10-py3.6-linux-x86_64.egg -P /home/colab/carla/PythonAPI/carla/dist/")

--2022-02-22 21:18:56--  https://github.com/MichaelBosello/carla-colab/raw/master/carla-0.9.10-py3.6-linux-x86_64.egg
Resolving github.com (github.com)... 13.114.40.48
Connecting to github.com (github.com)|13.114.40.48|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/MichaelBosello/carla-colab/master/carla-0.9.10-py3.6-linux-x86_64.egg [following]
--2022-02-22 21:18:57--  https://raw.githubusercontent.com/MichaelBosello/carla-colab/master/carla-0.9.10-py3.6-linux-x86_64.egg
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23488661 (22M) [application/octet-stream]
Saving to: ‘/home/colab/carla/PythonAPI/carla/dist/carla-0.9.10-py3.6-linux-x86_64.egg’

     0K .......... .......... .......... .....

Run CARLA

In [7]:
Call_SubprocessPopen("apt-get install libomp5")

Reading package lists...
Building dependency tree...
Reading state information...
The following packages were automatically installed and are no longer required:
  cuda-command-line-tools-10-0 cuda-command-line-tools-10-1
  cuda-command-line-tools-11-0 cuda-compiler-10-0 cuda-compiler-10-1
  cuda-compiler-11-0 cuda-cuobjdump-10-0 cuda-cuobjdump-10-1
  cuda-cuobjdump-11-0 cuda-cupti-10-0 cuda-cupti-10-1 cuda-cupti-11-0
  cuda-cupti-dev-11-0 cuda-documentation-10-0 cuda-documentation-10-1
  cuda-documentation-11-0 cuda-documentation-11-1 cuda-gdb-10-0 cuda-gdb-10-1
  cuda-gdb-11-0 cuda-gpu-library-advisor-10-0 cuda-gpu-library-advisor-10-1
  cuda-libraries-10-0 cuda-libraries-10-1 cuda-libraries-11-0
  cuda-memcheck-10-0 cuda-memcheck-10-1 cuda-memcheck-11-0 cuda-nsight-10-0
  cuda-nsight-10-1 cuda-nsight-11-0 cuda-nsight-11-1 cuda-nsight-compute-10-0
  cuda-nsight-compute-10-1 cuda-nsight-compute-11-0 cuda-nsight-compute-11-1
  cuda-nsight-systems-10-1 cuda-nsight-systems-11-0 cuda-nsig

In [8]:
# Call_SubprocessPopen("sudo -u colab DISPLAY=:1 vglrun /home/colab/carla/CarlaUE4.sh")

2/20/2022 10:50 PM:

TurboVNC: 

The global shader cache file '/home/colab/carla/Engine/GlobalShaderCache-GLSL_150_ES31.bin' is missing.

Your application is built to load COOKED content. No COOKED content was found; This usually means you did not cook content for this build. It also may indicate missing cooked data for a shader platform(e.g., OpenGL under Windows); Make sure your platform's packaging settings include this Targeted RHI.

Alternatively build and run the UNCOOKED version instead.

2/22/2022 3:41 PM: This error disappears for CARLA 0.9.10 instead of 0.9.13.

Client on Colab

In [9]:
Call_SubprocessPopen("sudo pip3 install pygame")
Call_SubprocessPopen("cd /home/colab/carla/PythonAPI/examples")
Call_SubprocessPopen("pwd")
Call_SubprocessPopen("python3 /home/colab/carla/PythonAPI/examples/manual_control.py")

Call_SubprocessPopen("sudo -u colab DISPLAY=:1 vglrun /home/colab/carla/CarlaUE4.sh")

python3: can't open file 'manual_control.py': [Errno 2] No such file or directory


KeyboardInterrupt: ignored